examples/ota-requestor-app/esp32/main/OTARequesterImpl.cpp

```diff
@@ -0,0 +1,282 @@
+        constexpr uint8_t kExampleProtocolsSupported =
+            EMBER_ZCL_OTA_DOWNLOAD_PROTOCOL_BDX_SYNCHRONOUS; // TODO: support this as a list once ember adds list support
+        const char locationBuf[] = { 'U', 'S' };
+        CharSpan exampleLocation(locationBuf);
+        constexpr bool kExampleClientCanConsent = false;
+        ByteSpan metadata;
+
+        err = cluster.QueryImage(successCallback, failureCallback, kExampleVendorId, kExampleProductId, kExampleHWVersion,
+                                 kExampleSoftwareVersion, kExampleProtocolsSupported, exampleLocation, kExampleClientCanConsent,
+                                 metadata);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(SoftwareUpdate, "QueryImage() failed: %s", chip::ErrorStr(err));
+        }
+        break;
+    }
+    case 2: { // ApplyUpdateRequest
+        chip::Callback::Cancelable * successCallback = mApplyUpdateResponseCallback.Cancel();
+        chip::Callback::Cancelable * failureCallback = mOnApplyUpdateRequestFailureCallback.Cancel();
+
+        constexpr uint32_t newVersion = 1;
+
+        err = cluster.ApplyUpdateRequest(successCallback, failureCallback, otaUpdateToken, newVersion);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(SoftwareUpdate, "ApplyUpdateRequest() failed: %s", chip::ErrorStr(err));
+        }
+        break;
+    }
+    case 3: { // NotifyUpdateApplied
+        constexpr uint32_t softwareVersion = 1;
+
+        err = cluster.NotifyUpdateApplied(nullptr, nullptr, otaUpdateToken, softwareVersion);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(SoftwareUpdate, "NotifyUpdateApplied() failed %s", chip::ErrorStr(err));
+        }
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void OnConnectionFailure(void * context, OperationalDeviceProxy * operationalDeviceProxy, CHIP_ERROR error)
+{
+    ChipLogError(SoftwareUpdate, "failed to connect to: %s", chip::ErrorStr(error));
+}
+
+void OnBlockReceived(void * context, chip::bdx::TransferSession::BlockData & blockdata)
+{
+    if (OTAUpdater::GetInstance().IsInProgress() == false)
+    {
+        OTAUpdater::GetInstance().Begin();
+    }
+    OTAUpdater::GetInstance().Write(reinterpret_cast<const void *>(blockdata.Data), blockdata.Length);
+}
+
+void OnTransferComplete(void * context)
+{
+    ESP_LOGI(TAG, "Transfer complete!");
+    OTAUpdater::GetInstance().End();
+}
+
+void OnTransferFailed(void * context)
+{
+    ESP_LOGI(TAG, "Transfer Failed");
+    OTAUpdater::GetInstance().Abort();
+}
+
+void ConnectToProvider(const char * ipAddress, uint32_t nodeId)
+{
+    // Explicitly calling UpdateAddress() should not be needed once OperationalDeviceProxy can resolve IP address from node ID and
+    NodeId providerNodeId = nodeId;
+    IPAddress ipAddr;
+    IPAddress::FromString(ipAddress, ipAddr);
+    PeerAddress addr = PeerAddress::UDP(ipAddr, CHIP_PORT);
```

Comment:
This CHIP_PORT is cut'n'paste. The port must be explicitly configured and should  come from oeprational discovery.

Reply:
This was taken from the linux example, it also uses the same `CHIP_PORT` and I think this code will be removed once `OperationalDeviceProxy` starts resolving IP address using fabric id and node id.